File: UPS_PHP_AuthCode_Sdk/src/AuthCodeService.php
<?php
namespace UpsPhpAuthCodeSdk;
require_once 'AuthCodeConstants.php';
require_once 'TokenInfo.php';
require_once 'UpsOauthResponse.php';
require_once 'HttpClient.php';

class AuthCodeService
{
    private $httpClient;
    public function __construct($httpClient) 
    {  
        $this->httpClient = $httpClient;
    }
    
    public function login($queryParams = null)
    {
        try {
            $url = $this->build_url_with_query_params(AuthCodeConstants::AUTHORIZE_URL, $queryParams);
            $httpHeaders = array(
                "Content-Type: application/x-www-form-urlencoded"
            );
            $response = $this->httpClient->post($url, $httpHeaders, null, 'GET');
            $response=json_decode($response);
            if ($response->status_code == 200 ) {
                $check = $this->login_response($response->redirect_url);
                return json_encode($check);
            } else {
                return json_encode($this->login_error_response($response->response));
            }
        } catch (Exception $e) {
            return $this->api_error_response(AuthCodeConstants::INTERNAL_SERVER_ERROR);
        }
    }
   
    public function get_access_token($client_id, $client_secret, $redirect_uri, $auth_code) {
        try {
            $body = array(
                "grant_type" => "authorization_code",
                "redirect_uri" => $redirect_uri,
                "code" => $auth_code
            );
            $response = $this->post_for_token_info(AuthCodeConstants::TOKEN_URL, $body, $client_id, $client_secret);
            return json_encode($response);
        } catch (Exception $e) {
            return $this->api_error_response(AuthCodeConstants::INTERNAL_SERVER_ERROR);
        }
    }

    public function post_for_token_info($url, $body, $clientId, $clientSecret) {
        try {
            $httpHeaders = array(
                "Content-Type: application/x-www-form-urlencoded",
                "Authorization: Basic " . base64_encode($clientId . ":" . $clientSecret)
            );
            $response = $this->httpClient->post($url, $httpHeaders, $body, 'POST');
            $response=json_decode($response);
            if ($response->status_code == 200) {
                return $this->api_response($response->response);
            } else if($response->error_number == 0) {
                return $this->api_error_response($response->response);
            } if($response->error_number > 0) {
                return $this->api_error_response(AuthCodeConstants::TIMED_OUT);
            }
        } catch (Exception $e) {
            return $this->api_error_response(AuthCodeConstants::INTERNAL_SERVER_ERROR);
        }
    }

    public function get_access_token_from_refresh_token($client_id, $client_secret, $refresh_token) {
        try {
            $body = array(
                "grant_type" => "refresh_token",
                "refresh_token" => $refresh_token
            );
            $response = $this->post_for_token_info(AuthCodeConstants::REFRESH_TOKEN_URL, $body, $client_id, $client_secret);
            return json_encode($response);
        } catch (Exception $e) {
            return $this->api_error_response(AuthCodeConstants::INTERNAL_SERVER_ERROR);
        }
    }
    public function build_url_with_query_params($base_url, $query_params) {
        if ($query_params == null) {
            return $base_url;
        }
        $query_string = http_build_query($query_params);
        return $base_url . "?" . $query_string;
    }

    public function api_response($response)
    {
        $json = json_decode($response);
        if (isset($json->access_token))
        {
            $token_info = array(
                "access_token" => $json->access_token,
                "client_id" => $json->client_id,
                "expires_in" => $json->expires_in,
                "issued_at" => $json->issued_at,
                "refresh_token" => $json->refresh_token,
                "refresh_token_expires_in" => $json->refresh_token_expires_in,
                "refresh_token_issued_at" => $json->refresh_token_issued_at,
                "refresh_token_status" => $json->refresh_token_status,
                "status" => $json->status,
                "token_type" => $json->token_type
            );
            return (new UpsOauthResponse($token_info, null))->to_dict();
        }
        return api_error_response($response);
    }

    public function api_error_response($json_data) {
        $error_response = json_decode(str_replace("'", '"', $json_data));
        //checking error object format
        if(isset($error_response->response->errors)){
            $errCode=$error_response->response->errors[0]->code;
            $errMsg=$error_response->response->errors[0]->message;
        }else{
            $errCode=$error_response->code;
            $errMsg=$error_response->message;
        }
        $api_error_info = new ErrorModel($errCode,$errMsg);
        return (new UpsOauthResponse(null, $api_error_info))->to_dict();
    }
   
    public function login_response($redirect_uri) {
        $login_info = new LoginInfo($redirect_uri);
        return (new UpsOauthResponse($login_info->to_dict(), null))->to_dict();
    }
   
    public function login_error_response($json_data) {
        $error_response = json_decode($json_data);
        $login_error_info = new ErrorModel(
            $error_response->response->errors[0]->code,
            $error_response->response->errors[0]->message);
        return (new UpsOauthResponse(null, $login_error_info))->to_dict();
    }

}
?>

File: UPS.Go.AuthCode.SDK/src/AuthCodeService.go
package authcode

import ( 
    "io"
    "net/http"
    "net/url"
    "strings"
    "time"
    "encoding/json"
    "net"
)

const (
    tokenUrl              = "https://onlinetools.ups.com/security/v1/oauth/token"
    refreshTokenUrl       = "https://onlinetools.ups.com/security/v1/oauth/refresh"
    loginUrl              = "https://onlinetools.ups.com/security/v1/oauth/authorize"
    timeoutDuration       = 30 * time.Second
    tlsHandshakeTimeout   = 10 * time.Second
    idleConnTimeout       = 10 * time.Second
    responseHeaderTimeout = 10 * time.Second
    expectContinueTimeout = 10 * time.Second
    timedOut = `{"response":{"errors":[{"code":"10500","message":"Request Timed out."}]}}`
    internalServerError = `{"response":{"errors":[{"code":"10500","message":"Internal server error"}]}}`
)

func setHttpClientTimeouts(httpClient *http.Client) *http.Client {
    if httpClient == nil {
        return &http.Client {
            Timeout: timeoutDuration,
            Transport: &http.Transport {
                TLSHandshakeTimeout: tlsHandshakeTimeout,
                IdleConnTimeout: idleConnTimeout,
                ResponseHeaderTimeout: responseHeaderTimeout,
                ExpectContinueTimeout: expectContinueTimeout,
            },
        }
    }

    httpClient.Timeout = timeoutDuration

    if transport, ok := httpClient.Transport.(*http.Transport); ok {
        transport.TLSHandshakeTimeout = tlsHandshakeTimeout
        transport.IdleConnTimeout = idleConnTimeout
        transport.ResponseHeaderTimeout = responseHeaderTimeout
        transport.ExpectContinueTimeout = expectContinueTimeout
    } else {
        httpClient.Transport = &http.Transport {
            TLSHandshakeTimeout: tlsHandshakeTimeout,
            IdleConnTimeout: idleConnTimeout,
            ResponseHeaderTimeout: responseHeaderTimeout,
            ExpectContinueTimeout: expectContinueTimeout,
        }
    }
    return httpClient
}

func Login(httpClient *http.Client, queryParams map[string]string ) (UpsOauthResponse) {
    var _httpClient *http.Client = setHttpClientTimeouts(httpClient)
        
    urlWithParams, err := url.Parse(loginUrl)
    if err != nil {
        return apiErrorResponse(err.Error())
    }

    query := urlWithParams.Query()
    for key, value := range queryParams {
        query.Set(key, value)
    }
    urlWithParams.RawQuery = query.Encode()

    req,err := http.NewRequest(http.MethodGet, urlWithParams.String(), nil)
    if err != nil {
        return apiErrorResponse(err.Error())
    }
        
    req.Header.Set("Content-Type", "application/x-www-form-urlencoded")    
    res, err := _httpClient.Do(req)
    if err != nil {
        if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
            return apiErrorResponse(timedOut)
        }
        return apiErrorResponse(err.Error())
    }

    defer res.Body.Close()    

    if !(res.StatusCode >= 200 && res.StatusCode <= 299) {
        response, err := io.ReadAll(res.Body)
        if err != nil {
            return apiErrorResponse(internalServerError)
        }
        return apiErrorResponse(string(response))
    }

    redirectUrl := res.Request.URL.String()

    redirectResponse := TokenInfo{
        RedirectURI: redirectUrl,
    }

    upsResponse := UpsOauthResponse{
        Response: redirectResponse,
        Error: "",
    }
    
    return upsResponse
}

func GetAccessToken(httpClient *http.Client, clientId string, clientSecret string, redirectUri string, authCode string) (UpsOauthResponse) {
    var _httpClient *http.Client = setHttpClientTimeouts(httpClient)   

    body := url.Values{}
    body.Set("grant_type", "authorization_code")
    body.Set("redirect_uri", redirectUri)
    body.Set("code", authCode)

    encodedData := body.Encode()

    req,err := http.NewRequest(http.MethodPost, tokenUrl, strings.NewReader(encodedData))
    if err != nil {        
        return apiErrorResponse(err.Error())
    }

    req.SetBasicAuth(clientId, clientSecret)
    req.Header.Set("Content-Type", "application/x-www-form-urlencoded")   

    res, err := _httpClient.Do(req)
    if err != nil {  
        if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
            return apiErrorResponse(timedOut)
        }
        return apiErrorResponse(err.Error())
    }

    defer res.Body.Close()
    
    response, err := io.ReadAll(res.Body)    
    if err != nil {
        return apiErrorResponse(internalServerError)
    }
    respString := string(response)
    var data TokenInfo
    errr := json.Unmarshal([]byte(respString), &data)
    if errr != nil {
        return apiErrorResponse(internalServerError)
    } 
    
    if !(res.StatusCode >= 200 && res.StatusCode <= 299) {
        return apiErrorResponse(string(response))
    }
    

    upsResponse := UpsOauthResponse{
        Response: data,
        Error: "",
    }
    return upsResponse
}

func GetAccessTokenFromRefreshToken(httpClient *http.Client, clientId string, clientSecret string, refreshToken string) (UpsOauthResponse) {
    var _httpClient *http.Client = setHttpClientTimeouts(httpClient)   
    
    if httpClient == nil {
        _httpClient = &http.Client {
            Timeout: timeoutDuration,
            Transport: &http.Transport {
                TLSHandshakeTimeout: tlsHandshakeTimeout,
                IdleConnTimeout: idleConnTimeout,
                ResponseHeaderTimeout: responseHeaderTimeout,
                ExpectContinueTimeout: expectContinueTimeout,
            },
        }

    } else {
        _httpClient = httpClient
    }

    body := url.Values{}
    body.Set("grant_type", "refresh_token")
    body.Set("refresh_token", refreshToken)

    encodedData := body.Encode()

    req,err := http.NewRequest(http.MethodPost, refreshTokenUrl, strings.NewReader(encodedData))
    if err != nil {        
        return apiErrorResponse(internalServerError)
    }

    req.SetBasicAuth(clientId, clientSecret)
    req.Header.Set("Content-Type", "application/x-www-form-urlencoded")   

    res, err := _httpClient.Do(req)
    if err != nil {        
        if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
            return apiErrorResponse(timedOut)
        }
        return apiErrorResponse(err.Error())
    }

    defer res.Body.Close()

    response, err := io.ReadAll(res.Body)

    if err != nil {        
        return apiErrorResponse(internalServerError)
    }
    
    respString := string(response)
    var data TokenInfo
    errr := json.Unmarshal([]byte(respString), &data)
    if errr != nil {
        return apiErrorResponse(internalServerError)
    } 

    if !(res.StatusCode >= 200 && res.StatusCode <= 299) {
        return apiErrorResponse(respString)
    }

    upsResponse := UpsOauthResponse{
        Response: data,
        Error: "",
    }

    return upsResponse
}

func apiErrorResponse(errorResponse string) UpsOauthResponse {    
    response := UpsOauthResponse{
        Response: TokenInfo{},
        Error: errorResponse,
    } 
    return response
}

type TokenInfo struct {
    Issued_at                string `json:"issued_at"`
    Token_type               string `json:"token_type"`
    Client_id                string `json:"client_id"`
    Access_token             string `json:"access_token"`
    Refresh_token            string `json:"refresh_token"`
    Refresh_token_issued_at  string `json:"refresh_token_issued_at"`
    Refresh_token_expires_in string `json:"refresh_token_expires_in"`
    Refresh_token_status     string `json:"refresh_token_status"`
    Expires_in               string `json:"expires_in"`
    Status                   string `json:"status"`
    RedirectURI              string `json:"redirect_uri"`
}


type LoginInfo struct {
     RedirectURL string `json:"redirect_url"`
}

type UpsOauthResponse struct {
	Response TokenInfo
	Error string
}
